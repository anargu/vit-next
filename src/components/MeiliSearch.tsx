import React from 'react';
import { InstantSearch, SearchBox, Hits, Highlight } from 'react-instantsearch-dom';
import { Hit } from 'react-instantsearch-core';
import { instantMeiliSearch } from '@meilisearch/instant-meilisearch';
import { VITResource } from '../core/entities';


const searchClient = instantMeiliSearch(
  process.env.MEILISEARCH_HOST_URL || "",
  process.env.MEILISEARCH_API_KEY ,
  {
    primaryKey: "id",
  }
);

type MeiliSearchBarProps = {
  ResultHitCard: React.ComponentType<{ hit: Hit<VITResource> }>,
  searchBarWrapperClassName?: string,
};

export const MeiliSearchBar = (
  { ResultHitCard, searchBarWrapperClassName } : MeiliSearchBarProps
) => (
  <div>
    <InstantSearch
      indexName="resources"
      searchClient={searchClient}
    >
      <div className={`${searchBarWrapperClassName || "mx-6"}`}><SearchBox /></div>
      
      <Hits hitComponent={ResultHitCard} />
    </InstantSearch>
  </div>
);

