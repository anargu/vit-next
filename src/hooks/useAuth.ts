import { onAuthStateChanged } from "firebase/auth";
import { useEffect, useState } from "react";
import { User } from "../core/entities";
import { auth } from "../firebase";

// Hook
export function useAuth () {
  const [isAuthenticated, setIsAuthenticated] = useState(false);
  const [authUser, setAuthUser] = useState<User | null>(null);

  useEffect(() => {
    if (isAuthenticated) return;

    const unSub = onAuthStateChanged(auth, (firebaseUser) => {
      if (!firebaseUser) return;

      const authUser : User = {
        id: firebaseUser!.uid,
        email: firebaseUser!.email!,
        displayName: firebaseUser!.displayName!,
      };

      setIsAuthenticated(!!authUser);
      setAuthUser(authUser);
    });

    return () => unSub && unSub();
  }, [isAuthenticated]);

  return {
    isAuthenticated,
    authUser,
  };
}


