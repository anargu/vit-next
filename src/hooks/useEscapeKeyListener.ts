import { useEffect } from 'react';

const useKeyUpListener = (keyString : string, callback: (arg0: KeyboardEvent) => void) => {
  useEffect(() => {
    const handleKeyUp = (event : KeyboardEvent) => {
      if (event.key === keyString) {
        callback(event);
      }
    };

    document.addEventListener('keyup', handleKeyUp);

    return () => {
      document.removeEventListener('keyup', handleKeyUp);
    };
  }, [callback]);
};

export default useKeyUpListener;

