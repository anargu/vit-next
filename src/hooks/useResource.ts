import { showDefaultNotification } from "../components/Notification/Notification";
import { Resource, VITResource } from "../core/entities";

export const useResource = () => {

  const share = (resource : Resource | VITResource) => {
    navigator.clipboard.writeText(resource?.url || "");

    showDefaultNotification("Link copied to Clipboard.");
  };

  return { share };
};
