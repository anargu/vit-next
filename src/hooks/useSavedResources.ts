import { useEffect, useState } from "react";
import { SAVED_LINK_KEY } from "../components/ResourceCard/ResourceCard";
import { VITResource } from "../core/entities";

export const useSavedResources = () => {

  const [savedResources, setSavedResources] = useState<null | VITResource[]>(null);

  useEffect(() => {
    if (!!savedResources) return;

    const rawData = localStorage.getItem(SAVED_LINK_KEY);
    if (!rawData) return setSavedResources([]);
    
    const savedPosts : VITResource[] | null = JSON.parse(rawData);
    if (!savedPosts) return setSavedResources([]);

    /* const _resources = savedPosts.map((post) => Resource.fromVITResource(post)); */
    
    setSavedResources(savedPosts);
  }, [savedResources]);

  return { savedResources };
};
