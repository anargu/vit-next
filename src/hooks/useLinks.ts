import { useEffect, useMemo, useState } from "react";
import useLocalStorageState from "use-local-storage-state";

import { useAuth } from "./useAuth";
import { DOMAIN_REGEX, SAVED_LINK_KEY } from "../core/constants";
import { fetchMetadataFromURL } from "../services/metadataService";
import { JSONResource, LinkPrivacy, Resource } from "../core/entities";
import { FirestoreStorageStrategy } from "../services/FirestoreStorageStrategy";
import { ga, ON_CLICK_SAVE_POST_EVENT, SAVE_POST_SUCESSFULLY_EVENT } from "../../lib/ga";

const firestoreDataManager = new FirestoreStorageStrategy();

export type UpdatePrivacySettingFn = (resourceId : string, privacy : LinkPrivacy) => Promise<void>;
export type SaveResourceFn = (resource : string | Resource) => Promise<void>;
export type DeleteResourceFn = (resourceId : string) => Promise<void>;

export interface useLinksOptions {
  listenSavedLinks?: boolean;
  listenFeedLinks?: boolean;
};

export function useLinks (opts: useLinksOptions = { listenSavedLinks: false, listenFeedLinks: false }) {

  const { authUser, isAuthenticated } = useAuth();

  const [_userLinks, _setUserLinks] = useState<Resource[] | null>([]);

  const [feedLinks, setFeedLinks] = useState<Resource[] | null>([]);

  // localLinks are saved links in LocalStorage 
  const [localLinks, setLocalLinks, { removeItem: deleteLocalLinks }] = useLocalStorageState<JSONResource[]>(SAVED_LINK_KEY, { defaultValue: [] });

  const userLinks = useMemo(() => {
    if (isAuthenticated) return _userLinks;

    return localLinks.map((localLink) => Resource.fromJSONResource(localLink));
  }, [isAuthenticated, localLinks, _userLinks]);

  // TODO: Implement this
  /* const initListeningLinksByUserId = (userId: string) => { */
  /*   const unSub = firestoreDataManager.listenLinksbyUserId(userId, (resources) => { */
  /*     _setUserLinks(resources); */
  /*   }); */
  /**/
  /*   return unSub; */
  /* }; */

  useEffect(() => {
    if (!authUser) return;
    if (!opts.listenSavedLinks) return;

    const unSub = firestoreDataManager.listenLinksbyUserId(authUser!.id, (resources) => {
      _setUserLinks(resources);
    });

    return () => {
      unSub?.();
    };
  }, [authUser?.id, opts.listenSavedLinks]);

  useEffect(() => {
    if (!opts.listenFeedLinks) return;

    const unSub = firestoreDataManager.listenFeedLinks((resources) => {
      setFeedLinks(resources);
    });

    return () => {
      unSub?.();
    };
  }, [opts.listenFeedLinks]);

  const insertLink = async (url : string) => {
    try {
      ga.event({ action: ON_CLICK_SAVE_POST_EVENT, params: {} });

      // Fetch Metadata of URL string
      let resourceMetadata = await fetchMetadataFromURL(url);

      // Save resource and its metadata
      const insertPromise = isAuthenticated
        ? firestoreDataManager.insert(resourceMetadata, authUser)
        : setLocalLinks([...localLinks, resourceMetadata.toJSON()]);

      await insertPromise;

      // Track analytics event
      const domain = url.match(DOMAIN_REGEX)?.[0];

      ga.event({
        action: SAVE_POST_SUCESSFULLY_EVENT,
        params: { domain: domain }
      });
    } catch (error : any) {
      console.error(error);

      throw error;
    }
  };

  const updateLinkPrivacy = async (resourceId: string, privacy: LinkPrivacy) => {
    try {
      if (!isAuthenticated) {
        throw Error("Operation cannot be done if user is not logged in.");
      }

      // User is logged in then update to DB.
      await firestoreDataManager.updateLinkPrivacy(resourceId, privacy);

    } catch (error : any) {
      console.error(error);

      throw error;
    }
  };

  const deleteLink = async (resourceId: string) => {
    try {
      if (!isAuthenticated) {
        setLocalLinks(localLinks.filter((_link) => _link.id !== resourceId));
        return;
      }

      await firestoreDataManager.delete(resourceId);
    } catch (error : any) {
      console.error(error);

      throw error;
    }
  };

  return {
    localLinks,
    deleteLocalLinks,

    userLinks,
    feedLinks,

    insertLink,
    deleteLink,
    updateLinkPrivacy,

    bulkInsertLinks: firestoreDataManager.bulkInsert,
  };
}
