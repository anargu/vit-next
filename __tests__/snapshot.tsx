import { render } from '@testing-library/react'
import Home from '@/pages/index'
import { useLinks } from '@/src/hooks/useLinks';
import { mockMeiliSearchComponent } from '@/src/components/MeiliSearch.mocks';

jest.mock("../src/components/MeiliSearch", () => ({
  __esModule: true,
  MeiliSearchBar: () => {
    return mockMeiliSearchComponent.MeiliSearchBar;
  },
}));

jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: () => {
    let pathname = "/";

    return {
      push: (newPath : string) => { pathname = newPath },
      pathname,
    };
  },
}));

jest.mock("@/src/hooks/useLinks", () => ({
  useLinks: jest.fn(),
}));

it('renders homepage unchanged', () => {
  (useLinks as jest.Mock).mockReturnValue({
    feedLinks: [],
    initListeningFeed: () => (() => {}),
  });

  const { container } = render(<Home />);
  expect(container).toMatchSnapshot();
})
