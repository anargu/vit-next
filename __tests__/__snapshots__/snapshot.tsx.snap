// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <div>
    <div
      class="grid grid-cols-[48px_48px_auto_70px] lg:grid-cols-[96px_96px_auto_70px] items-center h-[72px] mx-4"
    >
      <div
        class="text-base text-black-400 underline cursor-pointer text-center hover:text-gray-800"
      >
        Feed
      </div>
      <div
        class="text-base text-gray-400 cursor-pointer text-center hover:text-gray-800"
      >
        Saved
      </div>
      <div
        class="inline-block w-full"
      />
      <div
        class="inline-grid grid-cols-1 gap-6"
      >
        <div
          class="relative text-right"
        >
          <div>
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2728%27%20height=%2728%27/%3e"
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
          <div
            class="hidden right-0
        absolute z-20 min-w-[96px] text-base list-none bg-white rounded divide-y
        divide-gray-100 shadow dark:bg-gray-700 dark:divide-gray-600"
          >
            <ul
              aria-labelledby="dropdown"
              class="py-1"
            >
              <li>
                <a
                  class="block py-2 px-4 text-sm text-gray-700 hover:bg-gray-100 dark:hover:bg-gray-600 dark:text-gray-200 dark:hover:text-white"
                  href="/about"
                >
                  About us
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      class="grid grid-rows-[360px_min-content] min-h-[calc(100vh-72px)]"
    >
      <div
        class="text-stone-900 leading-tight text-4xl max-w-[80%] mx-6"
      >
        VIT, a curated feed of tons of interesting articles, news, tutorials... things.
      </div>
      <div>
        <form>
          <input
            type="search"
          />
          <button>
            o
          </button>
          <button>
            x
          </button>
        </form>
        <div>
          <ul>
            <li>
              1
            </li>
            <li>
              2
            </li>
            <li>
              3
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
